### PR TITLE
Allow platform admins to cancel broadcasts.

### PR DESCRIPTION
**Pull request by CrystalPea**
Do not allow platform admins to:
- create broadcasts
- approve broadcasts
- reject broadcasts

that is, unless they have a send_messages permission for a given service.

This is so platform admins have the minimum permissions necessary to cancel a broadcast that might have been sent out accidentally.

Pivotal ticket: https://www.pivotaltracker.com/story/show/177482418
API PR: https://github.com/alphagov/notifications-api/pull/3199

### CODE REVIEW

File: tests/app/main/views/test_broadcast.py [hunk 1]
```diff
@@ -158,6 +163,33 @@ def test_broadcast_pages_403_for_user_without_permission(
     )
 
 
+def test_cancel_broadcast_page_403_for_user_without_permission(
+    mocker,
+    client_request,
+    service_one,
+    active_user_view_permissions,
+    platform_admin_user_no_service_permissions,
```
Comment (benthorner) [inline tests/app/main/views/test_broadcast.py:171]: This isn't used in the test, so we can remove it.

**Review by benthorner — COMMENTED** [review]
Much safer 👍. I've left a couple of comments on the tests. 

When I tried this out locally I noticed buttons like "New alert", "Approve" and "Reject" are still there, but lead to a 403 page. I can see an argument for and against showing them. Is it intentional that we still show them?

File: tests/app/main/views/test_broadcast.py (hunk 1 above)
Comment (CrystalPea) [reply tests/app/main/views/test_broadcast.py:171]: good point 👍🏼

File: tests/app/main/views/test_broadcast.py [hunk 2]
```diff
@@ -2051,15 +2083,26 @@ def test_no_view_page_for_draft(
     )
 
 
+@pytest.mark.parametrize("user_is_platform_admin", [True, False])
 def test_cancel_broadcast(
     client_request,
     service_one,
+    service_two,
```
[the remark this answers is not in the document]
Comment (CrystalPea) [reply tests/app/main/views/test_broadcast.py]: good spot, it was me not spotting I didn't clean after myself (I had some issues with fixtures working like I wanted them to and was experimenting a bit before finally creating admin without service permissions fixture ). Cleaned up now ☺️

File: tests/conftest.py [hunk 3]
```diff
@@ -1108,6 +1108,31 @@ def platform_admin_user(fake_uuid):
     return user_data
 
 
+@pytest.fixture(scope='function')
+def platform_admin_user_no_service_permissions():
+    """
+    this fixture is for situations where we want to test that platform admin can access
+    an endpoint even though they have no explicit permissions for that service.
+    """
+    user_data = {'id': uuid4(),
```
[the remark this answers is not in the document]
Comment (CrystalPea) [reply tests/conftest.py:1117]: Hmm I agree it's good to keep our code dry, but I do have an objection: we usually call these fixtures in test definition, as arguments. It is not easy to pass anything into them at that point.

How do you propose we introduce them into tests alternatively if we refactor it? Would we then have to refactor hundreds of tests? If so, should we do it as a separate PR?

[the remark this answers is not in the document]
Comment (benthorner) [reply tests/conftest.py:1117]: So my suggestion was that we modify the existing [`create_platform_admin_user` method](https://github.com/alphagov/notifications-admin/blob/04d051df6b85cf596a7d6d0f28474b04673e420a/tests/conftest.py#L3976), passing in the things we want to change (I think that's just the service permissions), which we can default to an empty dict. 

In code that would look something like this:

```
@pytest.fixture
def platform_admin_user():
  create_platform_admin_user(permissions=...)

...

@pytest.fixture
def platform_admin_user_no_service_permissions
  create_platform_admin_user(permissions={})
```

This DRYs-up the knowledge about default data for a user into a single method (`create_platform_admin_user`), and we don't need to change any of the tests where the fixtures are used, so it's quite an easy refactor.

Does that help?

[the remark this answers is not in the document]
Comment (CrystalPea) [reply tests/conftest.py:1117]: Hey Ben, that helps a lot, makes a lot of sense now, thanks, I will refactor it like that ☺️

[the remark this answers is not in the document]
Comment (CrystalPea) [reply tests/conftest.py]: Ok done now :3

**Review by benthorner — APPROVED** [review]
Thanks for addressing all my initial comments. 

Somehow I've managed to analyse this PR much more thoroughly in the second review, and have a few more comments, some of them more major. Sorry about that. I always try hard to spot issues the first time round.

On the other hand, all of the issues I've spotted are in the tests, and aren't really that major: personally it's enough for me to see the `@user_has_permissions` handlers are in place, so I don't mind if the tests are all there.

I'd suggest fixing: 

- The comment about approved / rejected broadcasts.
- The broken "can't see a button" test for admins.

But I'm also happy for this to go in, especially since I won't be around to complete the review.

File: tests/app/main/views/test_broadcast.py [hunk 4]
```diff
@@ -357,16 +388,22 @@ def test_broadcast_dashboard(
...
     client_request,
     service_one,
     active_user_view_permissions,
+    platform_admin_user_no_service_permissions,
     mock_get_broadcast_messages,
     endpoint,
+    user_is_platform_admin
 ):
+    if user_is_platform_admin:
+        client_request.login(platform_admin_user_no_service_permissions)
```
[the remark this answers is not in the document]
Comment (CrystalPea) [reply tests/app/main/views/test_broadcast.py:405]: ahh, good spot, thanks!

File: tests/app/main/views/test_broadcast.py [hunk 5]
```diff
@@ -357,16 +388,22 @@ def test_broadcast_dashboard(
     )
 
 
+@pytest.mark.parametrize("user_is_platform_admin", [True, False])
 @pytest.mark.parametrize('endpoint', (
     '.broadcast_dashboard', '.broadcast_dashboard_previous', '.broadcast_dashboard_rejected',
```
[the remark this answers is not in the document]
Comment (CrystalPea) [reply tests/app/main/views/test_broadcast.py:439]: It's super weird we do not have any tests for `.view_broadcast` endpoint. I will make a separate ticket to add a set of tests for that endpoint.

[the remark this answers is not in the document]
Comment (CrystalPea) [reply tests/app/main/views/test_broadcast.py:439]: Ahh, there are tests, they are just under alternative endpoint names :o 
`.view_current_broadcast`
'.view_previous_broadcast`
`.view_rejected_broadcast`

**Pull request merged**